### PR TITLE
add support of bulk operations for ORM in ORACLE and SQLite backends

### PR DESCRIPTION
**Pull request by vyskubov**
Hello everyone,

I am excited to submit this pull request, which (re)introduces support for bulk operations in the SOCI library's ORM (Object-Relational Mapping) functionality. This work is based on the initial [groundwork](https://github.com/SOCI/soci/pull/168) laid by @abc100m, and I have built upon their contributions to enhance the capabilities of the library.

I would appreciate it if the community could review and provide feedback on the changes. Your insights and suggestions are valuable in ensuring the quality and effectiveness of this new feature.

Thank you for your attention and consideration. I look forward to collaborating with you all to further enhance the SOCI library.

Best regards,
Aleksey

### CODE REVIEW

**Review by vadz — COMMENTED** [review]
Thanks for the PR but I have a few questions just after the first very quick look at it:

1. Could you please try to minimize the number of unrelated changes? For examples, in several places `at(pos)` is replaced with `[pos]` which is not obviously right but even if it is, it would be much better to make it in a separate PR. There are also a lot of whitespace-only changes that really should be avoided.
2. Why are the new tests different for Oracle and SQLite? Is there some actual difference between the backends functionality and, if so, what is it? And if not, could we please have these tests in the common code?
3. And, of course, there are the the CI failures, could you please look at them?

I'll try to have a more detailed look at this later, but fixing (1) would be really helpful.

TIA!

**Comment by vadz** [comment]
>     1. I have reverted all the whitespace-only changes.

Thanks!

> Regarding the replacement of `at(pos)` with `[pos]`, these changes were already present in the original PR, so I simply included them as-is.

Sorry for not realizing this.

> However, I have now reverted these changes for the time being. It's worth mentioning that the usage of `at(pos)` and `[pos]` in this file is inconsistent, which could be addressed in a separate PR.

I agree that inconsistencies should be fixed but it's definitely better to do them in a separate PR. Thanks for fixing this up too!

**Comment by asmwarrior** [comment]
What is the status of this PR, I think the bulk adding is a very important feature for soci.

I have a similar discussion issue discussion here: [ORM: What is the correct way to bulk add a lot of rows? · Issue #1066 · SOCI/soci](https://github.com/SOCI/soci/issues/1066), in the issue, I have to use a for loop to add many rows, but I do need a simple way to do that.

Thanks.

**Comment by vyskubov** [comment]
I apologize for abandoning this PR. I would like to complete the work on it; however, I currently have more urgent tasks to attend to.

Furthermore, I do not have a readily available development environment on Windows and FreeBSD. Any assistance would be greatly appreciated.

Thank you.

File: tests/sqlite3/test-sqlite3.cpp [hunk 1]
```diff
@@ -397,6 +397,32 @@ struct longlong_table_creator : table_creator_base
...
+{
+    // we need to have an table that uses autoincrement to test this.
+    session sql(backEnd, connectString);
+
+    test4_table_creator tableCreator(sql);
+
+    {
+        std::vector<std::string> vec_name { "John", "James" };
+        sql << "insert into soci_test(name) values(:name)", use(vec_name);
+        assert(vec_name.size() == 2);
```
Comment (vadz) [inline tests/sqlite3/test-sqlite3.cpp:410]: All these `assert`s really need to be changed to `CHECK()`s.

File: tests/sqlite3/test-sqlite3.cpp [hunk 2]
```diff
@@ -676,6 +702,36 @@ class test_context : public test_context_base
...
+    session sql(backEnd, connectString);
+    table_creator_three one(sql);
+
+    sql << "insert into soci_test values('john', '(404)123-4567')";
+    sql << "insert into soci_test values('doe', '(404)123-4567')";
+
+    std::vector<PhonebookEntry> vec_orm(10);
+    sql << "select * from soci_test", into(vec_orm);
+
+    assert(vec_orm.size() == 2);
```
Comment (vadz) [inline tests/sqlite3/test-sqlite3.cpp:718]: Here too.

File: tests/sqlite3/test-sqlite3.cpp [hunk 3]
```diff
@@ -676,6 +702,36 @@ class test_context : public test_context_base
...
+
+    for (std::vector<PhonebookEntry>::iterator it = vec_orm.begin();
+         it != vec_orm.end();
+         ++it)
```
Comment (vadz) [inline tests/sqlite3/test-sqlite3.cpp:722]: Could use range-for loop here...

File: include/soci/row.h [hunk 4]
```diff
@@ -40,7 +46,7 @@ class SOCI_DECL column_properties
 class SOCI_DECL row
 {
 public:
-    row();
+    row(std::size_t bulk_size = 1);
```
Comment (vadz) [inline include/soci/row.h:49]:
```suggestion
    explicit row(std::size_t bulk_size = 1);
```

File: src/backends/sqlite3/vector-into-type.cpp [hunk 5]
```diff
@@ -132,8 +169,16 @@ void sqlite3_vector_into_type_backend::post_fetch(bool gotData, indicator * ind)
     int const endRow = static_cast<int>(statement_.dataCache_.size());
     for (int i = 0; i < endRow; ++i)
     {
-        sqlite3_column &col = statement_.dataCache_[i][position_-1];
-
+        std::shared_ptr<sqlite3_column> col1;
```
Comment (vadz) [inline src/backends/sqlite3/vector-into-type.cpp:172]: Why do we use a shared pointer here? It doesn't seem to be shared with anybody...

File: src/backends/sqlite3/vector-into-type.cpp [hunk 6]
```diff
@@ -132,8 +169,16 @@ void sqlite3_vector_into_type_backend::post_fetch(bool gotData, indicator * ind)
...
+        if (1 == endRow)
```
Comment (vadz) [inline src/backends/sqlite3/vector-into-type.cpp:173]: I wish there were a comment here explaining what is this test for. I can't understand what's going on here.

File: include/soci/sqlite3/soci-sqlite3.h [hunk 7]
```diff
@@ -60,6 +60,10 @@ class SOCI_SQLITE3_DECL sqlite3_soci_error : public soci_error
     int result_;
 };
 
+
+typedef std::map<std::string, data_type> sqlite3_data_type_map;
+sqlite3_data_type_map get_data_type_map();
+
```
Comment (vadz) [inline include/soci/sqlite3/soci-sqlite3.h:66]: There doesn't seem to be any reason to extract this here, is there?

File: include/soci/soci-platform.h [hunk 8]
```diff
@@ -64,6 +64,8 @@ namespace std {
     }
 }
 #endif // _MSC_VER < 1800
+#define strncasecmp _strnicmp
```
Comment (vadz) [inline include/soci/soci-platform.h:67]: This seems to be unused?

File: include/soci/values.h [hunk 9]
```diff
@@ -243,7 +246,7 @@ class SOCI_DECL values
     std::vector<details::standard_use_type *> uses_;
     std::map<details::use_type_base *, indicator *> unused_;
     std::vector<indicator *> indicators_;
-    std::map<std::string, std::size_t> index_;
+    std::map<std::string, std::size_t, CaseInsensitiveComparator> index_;
```
Comment (vadz) [inline include/soci/values.h:249]: Why did this need to be changed to be case-insensitive?

File: src/backends/oracle/statement.cpp [hunk 10]
```diff
@@ -298,6 +298,11 @@ void oracle_statement_backend::describe_column(int colNum, data_type &type,
             {
                 type = dt_double;
             }
+        }        
+        //create table T(t number)--> select t+0 from T; we get: scale=precision=0, size=22
+        else if (dbscale == 0 && dbprec == 0 && dbsize == 22)
+        {
+            type = dt_double;
```
Comment (vadz) [inline src/backends/oracle/statement.cpp:305]: This seems unrelated to the topic of this PR.

File: src/backends/sqlite3/statement.cpp [hunk 11]
```diff
@@ -19,12 +19,6 @@
 #pragma warning(disable:4355)
 #endif
 
-// This is used instead of tolower() just to avoid warnings about int to char
-// casts inside MSVS std::transform() implementation.
-char toLowerCh(char c) {
-    return static_cast<char>( std::tolower(c) );
-}
-
```
Comment (vadz) [inline src/backends/sqlite3/statement.cpp:27]: Removing this and inlining it looks a completely gratuitous change, was there really some reason for doing it?

File: src/backends/sqlite3/statement.cpp [hunk 12]
```diff
@@ -353,11 +350,7 @@ sqlite3_statement_backend::execute(int number)
 statement_backend::exec_fetch_result
 sqlite3_statement_backend::fetch(int number)
 {
-    if (hasVectorIntoElements_ || number == 0)
-        return load_rowset(number);
-    else
-        return load_one();
-
+    return (1 == number) ? load_one() : load_rowset(number);
```
Comment (vadz) [inline src/backends/sqlite3/statement.cpp:353]: It looks like `hasVectorIntoElements_` is not used at all now and should be removed?

File: include/soci/type-holder.h [hunk 13]
```diff
@@ -12,13 +12,29 @@
...
 #include <typeinfo>
+#include <vector>
 
 namespace soci
 {
 
 namespace details
 {
 
+// this class is for pass test [commontests.h, test12()],
```
Comment (vadz) [inline include/soci/type-holder.h:23]: This seems completely unrelated to the subject of this PR...

**Review by vadz — CHANGES_REQUESTED** [review]
Sorry, it doesn't seem like it can be merged in the current state, there are just too many unclear/unrelated things and commented out parts.

I also don't know if we really want to unconditionally replace scalar values with vectors, i.e. use vector of size 1 instead of simple values, it just looks like such a strange idea when we already have both scalar and vector versions.

Of course, maybe I just misunderstand things, but it's really hard to guess what is the intention/rationale here without any comments anywhere in sight.

**Comment by vadz** [comment]
It doesn't look like anything is going to happen here, so I have no choice but to close it, sorry.

**Pull request closed without merging**